### PR TITLE
Cancel Chat Requests (Ready)

### PR DESCRIPTION
**Pull request by hcsong213**
- [x] This branches off the tip of Max's misc commit and should be merged after it.

- Cancelation feature fully implemented for all mentors and mentees (not rigorously tested though)
- Reschedule not implemented
- Depends on bryan/modify-chat-req branch on backend

### CODE REVIEW

File: src/components/ReviewChats/ChatRequestMutators.tsx [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+          setClicked(false);
+        }}
+        onSecondaryButtonClick={() => {
+          setClicked(false);
+        }}
+        primaryButtonText="Submit"
+        secondaryButtonText="Close"
+      >
+        <Input
+          placeholder={"Reason for " + action + "ing (optional)"}
```
Comment (legitmaxwu) [inline src/components/ReviewChats/ChatRequestMutators.tsx:61]: I'm not sure how will this will generalize

**Review by legitmaxwu — CHANGES_REQUESTED** [review]
Awesome work--excited to see this merged soon so people can cancel chats.

Let some comments on functionality and style.

More general comments:
- The "reason for canceling" modal should make it more clear that you're writing to the other person
- Once a chat is cancelled, there's no way to view it. (It just disappears). Ideally this is not the case.

File: src/components/ReviewChats/ChatRequestMutators.tsx [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+import { Input, Text } from "../atomic";
+import CatchUnsavedChangesModal from "../CatchUnsavedChangesModal";
+import SelectOptionModal from "../SelectOptionModal";
+
+type ChatMutatorsProps = {
+  chatRequestId: string;
+  chatCanceled: boolean;
+};
+
+const ChatMutators = ({ chatRequestId, chatCanceled }: ChatMutatorsProps) => {
```
[the remark this answers is not in the document]
Comment (hcsong213) [reply src/components/ReviewChats/ChatRequestMutators.tsx:12]: It's supposed to eventually have both reschedule and chat; I didn't want to rename later

File: src/components/ReviewChats/ChatRequestMutators.tsx [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+  const [action, setAction] = useState("");
+  const [cancelReason, setCancelReason] = useState("");
+  const [modified, setModified] = useState(false);
+  const [cancelChatRequestMutation] = useCancelChatRequestMutation();
+
+  console.log(chatCanceled);
+  return (
+    <Fragment>
+      {/* This catch is super incomplete bc it doesn't prevent users from closing the modal mid-editting */}
+      <CatchUnsavedChangesModal unsavedChangesExist={modified === true} />
```
[the remark this answers is not in the document]
Comment (hcsong213) [reply src/components/ReviewChats/ChatRequestMutators.tsx:23]: sg

File: src/components/ReviewChats/ChatRequestMutators.tsx (hunk 1 above)
Comment (hcsong213) [reply src/components/ReviewChats/ChatRequestMutators.tsx:61]: It is supposed to be reused when rescheduling. In such a case, `action == "rechedule"`

Comment (hcsong213) [reply src/components/ReviewChats/ChatRequestMutators.tsx:61]: if that's kind of weird, it can be thrown out tho

**Review by legitmaxwu — APPROVED** [review]

**Pull request merged**